optapy-core/src/main/python/test/__init__.py

```diff
@@ -0,0 +1,609 @@
+from typing import Callable, Generic, List, Type, TypeVar, TYPE_CHECKING, overload, Union
+from jpype import JProxy
+
+from ..jpype_type_conversions import PythonBiFunction
+from ..optaplanner_java_interop import get_class
+from ..constraint_stream import PythonConstraintFactory, BytecodeTranslation
+
+if TYPE_CHECKING:
+    # These imports require a JVM to be running, so only import if type checking
+    from org.optaplanner.core.api.score.stream import Constraint, ConstraintFactory
+    from org.optaplanner.core.config.solver import SolverConfig
+    from org.optaplanner.core.api.score import Score
```

Comment:
What is type checking? Is that some sort of a phase that happens at a different time than at program run time?

Reply:
`TYPE_CHECKING` is `True` if the code being run by a type checker (an external program used by an IDE), `False` otherwise. It is used here to add type hints for an IDE; These imports will not necessarily be available when the module is imported (because the JVM did not start yet) and thus cab raise an ImportError if we import the package with them. Type stubs are generated for them, so a type checker knows about these classes, even though the implementation is not available.